src/main/java/io/cdap/plugin/gcp/dataplex/source/DataplexBatchSource.java

```diff
@@ -0,0 +1,209 @@
+      StructuredRecord transformed = outputSchema == null ?
+        transformer.transform((GenericData.Record) input.getValue()) :
+        transformer.transform((GenericData.Record) input.getValue(), outputSchema);
+      emitter.emit(transformed);
+    } else {
+      emitter.emit((StructuredRecord) input.getValue());
+    }
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    StageConfigurer configurer = pipelineConfigurer.getStageConfigurer();
+    FailureCollector collector = configurer.getFailureCollector();
+    if (!config.getConnection().canConnect() || config.getServiceAccountType() == null ||
+      (config.isServiceAccountFilePath() && config.autoServiceAccountUnavailable()) ||
+      (config.tryGetProject() == null)) {
+      // ValidatingInputFormat plugin setup is mandatory.Otherwise pipeline will fail at runtime in case of GCS entity.
+      config.setupValidatingInputFormat(pipelineConfigurer, collector, null);
+      return;
+    }
+    config.validateServiceAccount(collector);
+    entityBean = config.getAndValidateEntityConfiguration(collector, dataplexInterface);
+    if (entityBean == null) {
+      config.setupValidatingInputFormat(pipelineConfigurer, collector, null);
+      return;
+    }
+    if (entityBean.getSystem().equals(BIGQUERY_DATASET_ENTITY_TYPE)) {
+      getEntityValuesFromDataPathForBQEntities(entityBean.getDataPath());
+      config.validateBigQueryDataset(collector, datasetProject, dataset, tableId);
+      config.validateTable(collector, datasetProject, dataset, tableId);
+      Schema configuredSchema = getOutputSchemaForBQEntity(collector);
+      configurer.setOutputSchema(configuredSchema);
+    } else {
+      config.checkMetastoreForGCSEntity(dataplexInterface, collector);
+      config.setupValidatingInputFormat(pipelineConfigurer, collector, entityBean);
+    }
+  }
+
+  public void getEntityValuesFromDataPathForBQEntities(String dataPath) {
+    String[] entityValues = dataPath.split("/");
+    dataset = entityValues[entityValues.length - 3];
+    datasetProject = entityValues[1];
+    tableId = entityValues[entityValues.length - 1];
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+
+  }
+
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    super.initialize(context);
+    outputSchema = config.getSchema(context.getFailureCollector());
+  }
+
+
+  @Nullable
+  private Schema getOutputSchemaForBQEntity(FailureCollector collector) {
+    Schema outputSchema = config.getSchema(collector);
+    outputSchema = outputSchema == null ? getSchemaForBQEntity(collector) : outputSchema;
+    validateConfiguredSchemaForBQEntity(outputSchema, collector);
+    return outputSchema;
+  }
+
+  public Schema getSchemaForBQEntity(FailureCollector collector) {
```

Comment:
can we make this private?

Reply:
We will remove this method once Dataplex side, entity schema issue got fixed.